### ISSUE TITLE
IOException while reading MFT

### ISSUE THREAD

**Issue by jscarle** [DiscUtils/DiscUtils#227]
I've hit an [IOException](https://github.com/DiscUtils/DiscUtils/blob/08d7dd98cc4b10aad6bb5c7b3d23a61b64be77f2/Library/DiscUtils.Ntfs/MasterFileTable.cs#L344) while reading the the $MFT (BootRecord) of an NTFS volume of a VHDX from one of my virtual machines.

The issue occurs while reading the $MFT Data attributes. This particular VHDX is 3 TB in size and has hundreds of thousands of files. The $MFT has grown several times and is fragmented. In all likelyhood, there is nothing actually wrong with the MFT, its more likely that the code is hitting a particularity of the $MFT file itself. Microsoft clearly states that "when files are deleted, the associated FRSs are marked as free for reuse, but the total FRSs and associated MFT allocation remains." and that "Currently, only third-party defragmenters consolidate unused MFT FRS records and reclaim unused MFT allocated space." - https://docs.microsoft.com/en-us/troubleshoot/windows-server/backup-and-storage/disk-space-problems-on-ntfs-volumes#ntfs-master-file-table-mft-expansion

I've run every supported Microsoft tool on that volume and I cannot detect any corruption of the file system or the volume.

If I replace the call to IOException with a return null;, the flow continues and I am able to list and read both files and directories from that volume without an issue.

Below is a flow of the issue:

Reading MFT File...
    Reading Bootstrap File Record...
        Reading File Record 0...
            Reading NTFS Attributes...
                Reading StandardInformation Attribute...
                Reading FileName Attribute...
                Reading Data Attribute...
                Reading Data Attribute...
                Reading Data Attribute...
                Reading Data Attribute...
                Reading Data Attribute...
                    IOException: Attempt to get an MFT record with an old reference.
                Reading Data Attribute...
                    IOException: Attempt to get an MFT record with an old reference.
                Reading Data Attribute...
                    IOException: Attempt to get an MFT record with an old reference.
                Reading Data Attribute...
                    IOException: Attempt to get an MFT record with an old reference.
                Reading Data Attribute...
                    IOException: Attempt to get an MFT record with an old reference.
                Reading Data Attribute...
                    IOException: Attempt to get an MFT record with an old reference.
                Reading Data Attribute...
                    IOException: Attempt to get an MFT record with an old reference.
                Reading Bitmap Attribute...
                Reading Bitmap Attribute...

**Comment by jscarle** [comment]
I was able to narrow down the issue to the handling of the $20 NTFS Attribute:

![ProblematicMFT](https://user-images.githubusercontent.com/29762210/108640696-3553bb00-7469-11eb-89b0-64e40995de5f.PNG)

**Comment by EricZimmerman** [comment]
so this is fixed now? i occasionally see this happen and wondered why myself.

**Comment by jscarle** [comment]
No, its not. But the author of the project has effectively abandoned it.

**Comment by zivillian** [comment]
As far as I can tell there is no "the author" - many people have contributed, some moved on and there is still no decision how to move forward #260 

Instead of closing unsolved issued you could also try to solve solve it, open a PR and become a contributor or maintainer

**Comment by jscarle** [comment]
I've reopened the issue, hopefully one day someone will be able to fix this. Its not a very hard fix, its just an edge case with a difficult repro.

**Comment by zivillian** [comment]
can you share a test file to reproduce? this would make it much easier

**Comment by jscarle** [comment]
Unfortunately not.

**Comment by EricZimmerman** [comment]
If just returning null "fixes" things perhaps a check and warning for the old MFT record read can be displayed vs unhandled exception?

**Comment by jscarle** [comment]
My memory is a bit foggy as its been a few years, but the gist of what I remember is the following.

In NTFS, the MFT is stored in a special file named $Mft. The File Record Segment (FRS) for this file is stored as the first FRS in the MFT. It is always FRS 0. The FRS structure used to store the $Mft file is the same as any other FRS for any other regular file, and subject to the same limitations as a regular FRS for a regular file, except that it must always be # 0.

If you read Microsoft documentation on the NTFS MFT (https://learn.microsoft.com/en-us/windows/win32/devnotes/master-file-table), there is a single phrase that highlights the root cause of the issue.

"When there is no more space for storing attributes in the file record segment, additional file record segments are allocated and inserted in the first (or base) file record segment in an attribute called the attribute list."

As I recall, it is the presence of this attribute list that causes the IOException to occur, specifically the parsing of the entries (https://learn.microsoft.com/en-us/windows/win32/devnotes/attribute-list-entry).

Back then I had run a lot of tests and did a lot of investigation, I regret not having brought more details into this issue. I suppose I didn't imagine that'd it would be so long before it would be adressed.

The 0x20 $ATTRIBUTES_LIST is not normally present.

As a file becomes fragmented, new attributes are added the FRS, but as long as the number of attributes needed to reference each fragment can still be kept within the structure of a single FRS, then there's no need for the attributes list. However if enough fragments occur and the FRS becomes full, then the attributes list is created to store the references to the fragments of the file that exceed the capacity of the FRS.

This almost never occurs for two reasons.

Most regular files are regularly defragmented, therefore their FRS usually do not contain many attributes.

The number of attributes needed to reach the need to create an attributes list is ridiculously high.

For the $Mft file to contain an attributes list, the $Mft file must reach two conditions. It must references a ridiculously large amount of files that are created and deleted frequently which cause a ridiculously high amount of fragmentation (this is why it only occured on a large 3TB VHDX that if I recall correctly was an Exchange server). And the $Mft itself must be space constrained and fragment.

That's about what I remember.

**Comment by EricZimmerman** [comment]
I don't know about this. I think the attribute list is pretty common. The difference is whether it's resident or not. 

The data attribute (0x80) would be what tracks fragmentation. You'd just have more data runs if a file is fragmented. You'd need a really fragmented file to blow out what a typical record can hold. 

I have a fully featured MFT parser under my profile here or look at ericzimmermantools for an executable (mftecmd). Run it with the debug flag to see every attribute in the MFT. 

Of course if it's resident you'd see it in the MFT entry directly but if it's non resident you'd need the disk image to find it, just like a directory index. 

Some additional details are here

https://github.com/libyal/libfsntfs/blob/main/documentation/New%20Technologies%20File%20System%20(NTFS).asciidoc#63-the-attribute-list-attribute


It should also be noted that when a file record exceeds the space of a single MFT record (almost always 1024 bytes) an extension record is used to store MORE attributes. This is just another FILE record somewhere else in the MFT.

This is where most MFT parsers that just deal with an MFT file fall down. Extension records have to be handled differently than a normal FILE record. 

I have hundreds of mfts and can check for how frequent the attribute list is but from my recollection it was pretty common. 

For this error, I see it very rarely in 100 s of thousands of uses but it's super annoying when it does happen.

**Comment by jscarle** [comment]
As I said, it was a few years ago, so the details are foggy. I just remember that it centered around the $Mft being fragmented, it containing millions of files, and the attributes list referencing FRSes that had been deleted I think.

**Comment by LTRData** [comment]
I think it is safe to just return null from `GetRecord()` in these cases. When I investigated it a few years ago I could not find any issues with doing that. Unfortunately, that change was only made to my fork and not to this repo. But I could probably make a PR for it here as well.

**Comment by EricZimmerman** [comment]
It would fix the fringe issue for sure. If your fork has a nugget package I
can switch to that tho.

**Comment by LTRData** [comment]
Yes, my fork is at http://github.com/LTRData/DiscUtils and nuget packages have the same names as here but with `LTRData.` prefix, for example `LTRData.DiscUtils.Ntfs`. https://www.nuget.org/packages/LTRData.DiscUtils

It should be said that there are a few breaking changes here and there, mainly where it has been required for performance improvements. Things like returning enumerators instead of arrays and similar. But it is typically not that difficult to make necessary changes for it.
